docs/design/websockets-watches-proposal.md

```diff
@@ -0,0 +1,184 @@
+
+* The `serveWatch()` in `pkg/apiserver/resthandler.go` parses the request and creates a WebSocket
+server to handles multiple watches. 
+
+* The WebSocket server in `pkg/apiserver/watch.go` creates a slice of watches for resources.
+
+```
+type WatchServer struct {
+watching []watch.Interface
+.
+.
+}
+```
+
+* The WebSocket handler HandleWS in `pkg/apiserver/watch.go` handles all watches.
+
+#### Validation
+
+The resthandler validates that requested resources are indeed one of the existing kube resources.
+
+#### Limiting number of resources per request
+
+This solution also requires limiting number of resources to watch per request, as allowing
+number of resources without upper bound would lead to an increased amount of response data that
+might overwhelm the connection.
+
+```
+type WatchServer struct {
+watching []watch.Interface
+.
+.
+maxAllowedRequest int32
+}
+```
+
+Where `maxAllowedRequest` could be something `5` or `10` (or something else).
+
+### Multiple watch requests at different times over a WebSocket Connection
+
+In this solution, the expected steps are:
+
+1. A client opens a WebSocket connection to a Kubernetes API server.
+
+2. The client then sends a watch request for a particular resource over the connection.
+
+3. The api server responds to the request.
+
+4. The client then may open another watch request on the same connection for a different resource
+while the previous watch is still open over the same connection.
+
+5. The api server responds to the another watch request over the same connection.
+
+6. The steps 4 and 5 are repeated as needed.
+
+7. The server may close the connection if all watches are finished for whatever reason or
+the client closes the connection.
+
+This solution is different from the previous one because watch requests are not sent
+in bulk in the begining as in the previous solution.
+
+#### Implementation details
+
+To-Do
+
+## Open Issues
+
+### Impact of more watches on etcd
+
+In kubernetes, when a client sends a watch request to kubernetes apiserver, the apiserver opens an watch on etcd
+for each watch request. With the growing number of resource types in kubernetes, and the need for watching more
+resources simultaneously, it is going to put even more burden on etcd's performance and kubernetes performance
+in general. This performance impact on etcd needs to be addressed and there is a proposal [apiserver-watch](https://github.com/kubernetes/kubernetes/blob/release-1.4/docs/proposals/apiserver-watch.md)
+that deals with it.
+
+Although the above solution requires opening more watches simultaneously, the current/short plan for now is to not
+the addresses the performance impact on etcd (if any) as part of this proposal.
+
+### Security Considerations
+
+What client/user is allowed to request watches on what resources.
+
+## Roadmap
+
+1. The firt solution above for kube 1.5.
```

Comment:
i would flip the solution order since our UI needs the websocket pattern before the generic pattern.